proxy/agent.go

```diff
@@ -99,28 +97,37 @@ type Retry struct {
 	// delay between reconciliation adjusted by retries
 	delay time.Duration
 
-	// default delay is a long duration - reconciliation is a no-op in the regular case.
-	// For permanent errors, reconciliation is still attempted once after the default delay.
-	defaultDelay time.Duration
-
 	// number of times to attempts left to retry applying the latest desired configuration
 	budget int
 
-	// maximum number of retries
-	maxRetries int
+	// DefaultDelay is a default long duration for regular reconciliation (a no-op if the config is unchanged)
+	DefaultDelay time.Duration
+
+	// MaxRetries is the maximum number of retries
+	MaxRetries int
 
-	// delay between the first restart, from then on it is multiplied by a factor of 2
-	// for each subsequent retry
-	initialInterval time.Duration
+	// InitialInterval is the delay between the first restart, from then on it is
+	// multiplied by a factor of 2 for each subsequent retry
+	InitialInterval time.Duration
+}
+
+// reset retry budget
+func (r *Retry) reset() {
+	r.delay = r.DefaultDelay
+	r.budget = r.MaxRetries
 }
 
 // Proxy defines command interface for a proxy
 type Proxy struct {
-	// Run command for a config, epoch and abort channel
+	// Run command for a config, epoch, and abort channel
 	Run func(interface{}, int, <-chan error) error
 
-	// Cleanup command
+	// Cleanup command for an epoch
 	Cleanup func(int)
+
+	// Panic command is invoked when all retries to start the proxy fail
+	// just before the agent terminating
+	Panic func()
```

Comment:
I think we need to wave a red flag here since it's a very bad error case. Any proposal?